libpromises/evalfunction.c

```diff
@@ -1695,6 +1701,265 @@ static FnCallResult FnCallSplayClass(EvalContext *ctx,
 
 /*********************************************************************/
 
+#ifdef HAVE_LIBCURL
+struct _curl_userdata
+{
+    const FnCall *fp;
+    const char *desc;
+    size_t max_size;
+    Buffer* content;
+};
+
+static size_t cfengine_curl_write_callback(char *ptr, size_t size, size_t nmemb, void *userdata)
+{
+    struct _curl_userdata *options = (struct _curl_userdata*) userdata;
+    unsigned int old = BufferSize(options->content);
+    size_t requested = size*nmemb;
+    size_t granted = requested;
+
+    if (old + requested > options->max_size)
+    {
+        granted = options->max_size - old;
+        Log(LOG_LEVEL_VERBOSE, "%s: while receiving %s, current %u + requested %zd bytes would be over the maximum %zd; only accepting %zd bytes", options->fp->name, options->desc, old, requested, options->max_size, granted);
+    }
+
+    BufferAppend(options->content, ptr, granted);
+
+    // `written` is actually (BufferSize(options->content) - old) but
+    // libcurl doesn't like that
+    size_t written = requested;
+
+    // extra caution
+    BufferTrimToMaxLength(options->content, options->max_size);
+    return written;
+}
+
+static void CurlCleanup()
+{
+    if (NULL == CURL_CACHE)
+    {
+        JsonElement *temp = CURL_CACHE;
+        CURL_CACHE = NULL;
+        JsonDestroy(temp);
+    }
+
+    if (CURL_INITIALIZED)
+    {
+        curl_global_cleanup();
+        CURL_INITIALIZED = false;
+    }
+
+}
+#endif
+
+static FnCallResult FnCallUrlGet(ARG_UNUSED EvalContext *ctx,
+                                 ARG_UNUSED const Policy *policy,
+                                 const FnCall *fp,
+                                 const Rlist *finalargs)
+{
+#ifdef HAVE_LIBCURL
+    char *url = RlistScalarValue(finalargs);
+    VarRef *ref = ResolveAndQualifyVarName(fp, RlistScalarValue(finalargs->next));
+    if (!ref)
+    {
+        return FnFailure();
+    }
+
+    JsonElement *options = VarRefValueToJson(ctx, fp, ref, NULL, 0);
+    VarRefDestroy(ref);
+    if (NULL == options)
+    {
+        return FnFailure();
+    }
+
+    if (JsonGetElementType(options) != JSON_ELEMENT_TYPE_CONTAINER ||
+        JsonGetContainerType(options) != JSON_CONTAINER_TYPE_OBJECT)
+    {
+        JsonDestroy(options);
+        return FnFailure();
+    }
+
+    Writer *cache_w = StringWriter();
+    WriterWriteF(cache_w, "url = %s; options = ", url);
+    JsonWriteCompact(cache_w, options);
+
+    if (NULL == CURL_CACHE)
+    {
+        CURL_CACHE = JsonObjectCreate(10);
+    }
+    atexit(&CurlCleanup);
```

Comment:
This should be called inside the if to avoid registering it every time someone calls ``url_get``, or?